### PR TITLE
Integration Bus

### PR DESCRIPTION
**Pull request by armiol**
This PR introduces `IntegrationBus` and ability to exchange messages between `BoundedContext`s.

Highlights:

* `IntegrationBus` along with `TransportFactory` are introduced and represent a mechanism of inter-`BoundedContext` communication. Each `BoundedContext` may produce events and rejections in which other contexts are interested (the latter are **downstream** in this case). The messages of interest are brought to their subscribers/reactors via a customizable `Transport`, inspired by [Publish-Subscribe](http://www.enterpriseintegrationpatterns.com/patterns/messaging/PublishSubscribeChannel.html) messaging systems.

* `React(external = true)` and `Subscribe(external = true)` now allow to process events and rejections that are fired by another bounded context. Handler methods marked `external` play an **anticorruption layer** role (see ACL pattern in "Domain-Driven Design" by Eric Evans and "Implementing Domain-Driven Design" by Vaughn Vernon) .

* `TransportFactory` and `MessageChannel` concepts are introduced. There is also a naïve in-memory transport implementation for test purposes.

In addition, `BoundedContextId` is introduced to identify bounded contexts. It is an improvement to a `String` used for identification previously.

**IntegrationBus**

This bus is available as a part of single `BoundedContext`. In a multi-component environment messages may travel across components from one bounded context to another. 

`IntegrationBus` is always based upon some transport (see more on that below), that delivers the messages from and to it. For several bounded contexts to communicate, their integration buses have to share the transport. Typically that would be a single messaging broker.

The messages from external components received by the `IntegrationBus` instance via the transport are propagated into the bounded context. They are dispatched to the subscribers and reactors marked with `external = true` on per-message-type basis.

`IntegrationBus` is also responsible for publishing the messages, born within the current `BoundedContext`, to external collaborators. To do that properly, the bus listens to a special document message called `RequestForExternalMessages`, that describes the needs of other parties.

E.g. bounded context `Projects` has the external event handler method in the projection as follows:

```java
public class ProjectListView extends Projection<...> {

    @Subscribe(external = true)
    public void on(UserDeleted event) {
        // Remove the projects that belong to this user.
        // ...
    }

    // ...
}
```
Upon a registration of the corresponding repository the  integration bus of `Projects` context sends out a `RequestForExternalMessages` saying that an `Event` of `UserDeleted` is needed from other parties. 

Let's say the second bounded context is `Users`. Its integration bus will receive the `RequestForExternalMessages` request sent out by `Projects`. To handle it properly it will create a bridge between `Users`'s event bus (which may eventually be transmitting `UserDeleted` event) and an external transport. Once `UserDeleted` is published locally in `Users` context, it will be received by this bridge (as well as other local dispatchers), and published to the external transport.

The integration bus of `Projects` context will receive the `UserDeleted` external message. The event will be dispatched to the external event handler of `ProjectListView` projection.

**TransportFactory and MessageChannel**

This PR introduces an initial implementation of [Publish-Subscribe Channel](http://www.enterpriseintegrationpatterns.com/patterns/messaging/PublishSubscribeChannel.html) pattern. 

The `TransportFactory` allows to create channels for outgoing messages (`Publisher`s) and incoming messages (`Subscriber`s).  Each channel serves messages of a single type.

In scope of this PR an in-memory implementation is provided, which is only suitable for the publishers and subscribers located within the same JVM. In the nearest future a Google Cloud-based implementation will be added to Spine.

The framework version is set to `0.9.65-SNAPSHOT`.

To be delivered in the following PRs:

* Google PubSub-based implementation of `Transport` along with proper integration tests.
* RabbitMQ-based implementation of `Transport` along with proper integration tests.

### CODE REVIEW

**Comment by armiol** [comment]
@alexander-yevsyukov PTAL.

**Comment by alexander-yevsyukov** [comment]
> This bus is available as a part of single `BoundedContext`

How do we connect an Integration Bus within BC1 to an Integration Context within BC2 then?

**Comment by armiol** [comment]
@alexander-yevsyukov 

> How do we connect an Integration Bus within BC1 to an Integration Context within BC2 then?

They need to connect by transport. E.g. have the same PubSub (or RabbitMQ, Kafka etc) broker. This is done by [configuring the `TransportFactory` that is supplied to `IntegrationBus.Builder`](https://github.com/SpineEventEngine/core-java/blob/integration-bus/server/src/test/java/io/spine/server/integration/given/IntegrationBusTestEnv.java#L112).

Originally my intention was to have an `IntegrationBus` inside a `BoundedContext` and some sort of `IntegrationService` external to bounded contexts — as a rendezvous point. However this idea did not work — it appeared too complex to design with no single point of failure.

With the current approach bounded contexts know nothing about each other, but are able to interconnect via transport channels in a shared space.

File: server/src/main/java/io/spine/server/projection/ProjectionRepository.java [hunk 1]
```diff
@@ -229,4 +234,31 @@ protected P findOrCreate(I id) {
...
+
+        @Override
+        public Set<ExternalMessageClass> getMessageClasses() {
+            final Set<EventClass> eventClasses = projectionClass().getExternalEventSubscriptions();
+            return ExternalMessageClass.fromEventClasses(eventClasses);
+        }
+
+        @Override
+        public void onError(ExternalMessageEnvelope envelope, RuntimeException exception) {
+
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/projection/ProjectionRepository.java]: Please remove the empty line.

File: server/src/main/java/io/spine/server/rejection/DelegatingRejectionDispatcher.java [hunk 2]
```diff
@@ -82,4 +98,42 @@ public String toString() {
...
+            public Set<I> dispatch(ExternalMessageEnvelope envelope) {
+                final ExternalMessage externalMessage = envelope.getOuterObject();
+                final Rejection rejection = unpack(externalMessage.getOriginalMessage());
+                final Set<I> ids = delegate.dispatchRejection(RejectionEnvelope.of(rejection));
+                return ids;
+            }
+
+            @Override
+            public void onError(ExternalMessageEnvelope envelope, RuntimeException exception) {
+
```
Comment (alexander-yevsyukov) [inline server/src/main/java/io/spine/server/rejection/DelegatingRejectionDispatcher.java]: Please remove the empty line.

File: server/src/main/java/io/spine/server/integration/IntegrationBus.java [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+ * <p>The messages from external components received by the {@code IntegrationBus} instance
+ * via the transport are propagated into the bounded context. They are dispatched
+ * to the subscribers and reactors marked with {@code external = true} on per-message-type basis.
+ *
+ * {@code IntegrationBus} is also responsible for publishing the messages,
+ * born within the current `BoundedContext`, to external collaborators. To do that properly,
+ * the bus listens to a special document message called {@linkplain RequestForExternalMessages},
+ * that describes the needs of other parties.
+ *
+ * <p>An example.
```
[the remark this answers is not in the document]
Comment (armiol) [reply server/src/main/java/io/spine/server/integration/IntegrationBus.java]: Done, but used `<h3>`. `<h1>` is too much.

File: server/src/main/java/io/spine/server/integration/LocalEventSubscriber.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * A subscriber to local {@code EventBus}, which publishes each matching message to
+ * a remote channel.
+ *
+ * <p>The messages to subscribe are those that are required by external application components
+ * at this moment; their set is determined by the {@linkplain RequestForExternalMessages
+ * configuration messages}, received by this instance of {@code IntegrationBus}.
+ */
+final class LocalEventSubscriber extends EventSubscriber {
```
[the remark this answers is not in the document]
Comment (armiol) [reply server/src/main/java/io/spine/server/integration/LocalEventSubscriber.java]: Renamed to `DomesticEventPublisher`.

File: server/src/main/java/io/spine/server/integration/LocalRejectionSubscriber.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+ * a remote channel.
+ *
+ * <p>The messages to subscribe are those that are required by external application components
+ * at this moment; their set is determined by the {@linkplain RequestForExternalMessages
+ * configuration messages}, received by this instance of {@code IntegrationBus}.
+ */
+final class LocalRejectionSubscriber extends RejectionSubscriber {
+
+    private final BoundedContextId boundedContextId;
+    private final TransportFactory.PublisherHub publisherHub;
```
[the remark this answers is not in the document]
Comment (armiol) [reply server/src/main/java/io/spine/server/integration/LocalRejectionSubscriber.java]: Renamed to `DomesticRejectionPublisher`.

**Comment by armiol** [comment]
@alexander-yevsyukov comments are addressed. PTAL again.

**Pull request merged**